ebpf tests fail due to missing bpf_endian.h

I see 17-18 ebpf tests fail saying “bfp_endian.h” not found for an include file in p4c/backends/ebpf/runtime/

I found the file via a google search and put it in my p4c/backends/ebpf/runtime/ and then then tests pass.

Failure log is included below when I run this test:

`$ ebpf/testdata/p4_16_samples/test_ebpf.p4.test -v`

```bash
*** Error 2: Failed to build the filter:
In file included from /home/hemant/p4c-exit-work/backends/ebpf/runtime/ebpf_test.h:29:0,
                 from /home/hemant/p4c-exit-work/backends/ebpf/runtime/ebpf_runtime_test.h:27,
                 from <command-line>:0:
/home/hemant/p4c-exit-work/backends/ebpf/runtime/ebpf_common.h:26:60: fatal error: bpf_endian.h: No such file or directory
compilation terminated.
make: *** [/home/hemant/p4c-exit-work/build/tmp40fa8r3v/build/ebpf_registry.o] Error 1
```

--- a/backends/ebpf/build_libbpf
+++ b/backends/ebpf/build_libbpf
@@ -39,9 +39,10 @@ def main():
     libbpf_target_dir = "%s/runtime" % file_dir
     # Download submodule if it has not been initialized yet
     git_cmd = "git submodule update --init %s" % libbpf_dir
-    err = testutils.run_timeout(verbose=VERBOSE, args=git_cmd,
-                                timeout=20, outputs=outputs,
-                                errmsg="Could not initialize libbpf submodule")
+    # We ignore the error that is thrown here
+    testutils.run_timeout(verbose=VERBOSE, args=git_cmd,
+                          timeout=20, outputs=outputs,
+                          errmsg="Could not initialize libbpf submodule")
     # Create the libbpf build directory
     create_dir(libbpf_build_dir)
     # Build libbpf

--- a/backends/ebpf/runtime/ebpf_common.h
+++ b/backends/ebpf/runtime/ebpf_common.h
@@ -23,8 +23,6 @@ limitations under the License.
 #include <stdbool.h>        // true and false
 #include <linux/types.h>    // u8, u16, u32, u64
 
-#include "bpf_endian.h" // definitions for bpf_ntohs etc...
-
 typedef signed char s8;
 typedef unsigned char u8;
 typedef signed short s16;
@@ -34,17 +32,6 @@ typedef unsigned int u32;
 typedef signed long long s64;
 typedef unsigned long long u64;
 
-
-#undef htonl
-#undef htons
-#define htons(d) __bpf_htons(d)
-#define htonl(d) __bpf_ntohl(d)
-#define htonll(d) __bpf_constant_be64_to_cpu(d)
-
-#define load_byte(data, b) (*(((u8*)(data)) + (b)))
-#define load_half(data, b) bpf_ntohs(*(u16 *)((u8*)(data) + (b)))
-#define load_word(data, b) bpf_ntohl(*(u32 *)((u8*)(data) + (b)))
-#define load_dword(data, b) bpf_cpu_to_be64(*(u64 *)((u8*)(data) + (b)))
 
 /*
  * Helper function.

--- a/backends/ebpf/runtime/ebpf_kernel.h
+++ b/backends/ebpf/runtime/ebpf_kernel.h
@@ -25,6 +25,20 @@ limitations under the License.
 
 #include "ebpf_common.h"
 
+#include "bpf_endian.h" // definitions for bpf_ntohs etc...
+
+#undef htonl
+#undef htons
+#define htons(d) bpf_htons(d)
+#define htonl(d) bpf_htonl(d)
+#define htonll(d) bpf_cpu_to_be64(d)
+#define ntohll(x) bpf_be64_to_cpu(x)
+
+#define load_byte(data, b) (*(((u8*)(data)) + (b)))
+#define load_half(data, b) bpf_ntohs(*(u16 *)((u8*)(data) + (b)))
+#define load_word(data, b) bpf_ntohl(*(u32 *)((u8*)(data) + (b)))
+#define load_dword(data, b) bpf_be64_to_cpu(*(u64 *)((u8*)(data) + (b)))
+
 
 /* If we operate in user space we only need to include bpf.h and
  * define the userspace API macros.

--- a/backends/ebpf/runtime/ebpf_test.h
+++ b/backends/ebpf/runtime/ebpf_test.h
@@ -28,6 +28,26 @@ limitations under the License.
 #include "ebpf_registry.h"
 #include "ebpf_common.h"
 
+#include <endian.h>
+
+/* define some byte order conversions, these mimic bpf_endian.h */
+#define htonll(x) htobe64(x)
+#define ntohll(x) be64toh(x)
+
+#define bpf_htons(x) htobe16(x)
+#define bpf_ntohs(x) be16toh(x)
+#define bpf_htonl(x) htobe32(x)
+#define bpf_ntohl(x) be32toh(x)
+#define bpf_cpu_to_be64(x) htobe64(x)
+#define bpf_be64_to_cpu(x) be64toh(x)
+
+#define load_byte(data, b) (*(((u8*)(data)) + (b)))
+#define load_half(data, b) bpf_ntohs(*(u16 *)((u8*)(data) + (b)))
+#define load_word(data, b) bpf_ntohl(*(u32 *)((u8*)(data) + (b)))
+#define load_dword(data, b) bpf_be64_to_cpu(*(u64 *)((u8*)(data) + (b)))
+
+
+
 #define bpf_printk(fmt, ...)                                            \
                 ({                                                      \
                         char ____fmt[] = fmt;                           \

--- a/testdata/extern_modules/extern-checksum-ebpf.c
+++ b/testdata/extern_modules/extern-checksum-ebpf.c
@@ -16,14 +16,14 @@
 static inline u8 verify_ipv4_checksum(const struct IPv4_h iphdr)
 {
     u8 correct = 0;
-    u32 checksum = __bpf_htons(((u16) iphdr.version << 12) | ((u16) iphdr.ihl << 8) | (u16) iphdr.diffserv);
-    checksum += __bpf_htons(iphdr.totalLen);
-    checksum += __bpf_htons(iphdr.identification);
-    checksum += __bpf_htons(((u16) iphdr.flags << 13) | iphdr.fragOffset);
-    checksum += __bpf_htons(((u16) iphdr.ttl << 8) | (u16) iphdr.protocol);
-    checksum += __bpf_htons(iphdr.hdrChecksum);
-    u32 srcAddr = __bpf_ntohl(iphdr.srcAddr);
-    u32 dstAddr = __bpf_ntohl(iphdr.dstAddr);
+    u32 checksum = bpf_htons(((u16) iphdr.version << 12) | ((u16) iphdr.ihl << 8) | (u16) iphdr.diffserv);
+    checksum += bpf_htons(iphdr.totalLen);
+    checksum += bpf_htons(iphdr.identification);
+    checksum += bpf_htons(((u16) iphdr.flags << 13) | iphdr.fragOffset);
+    checksum += bpf_htons(((u16) iphdr.ttl << 8) | (u16) iphdr.protocol);
+    checksum += bpf_htons(iphdr.hdrChecksum);
+    u32 srcAddr = bpf_ntohl(iphdr.srcAddr);
+    u32 dstAddr = bpf_ntohl(iphdr.dstAddr);
     checksum += (srcAddr >> 16) + (u16) srcAddr;
     checksum += (dstAddr >> 16) + (u16) dstAddr;
 

